sharedbuffers/mapped_struct.py

```diff
@@ -333,14 +335,185 @@ class mapped_dict(dict):
+        return proxied_dict.pack_into(obj, buf, offs, idmap, implicit_offs)
 
     @classmethod
     def unpack_from(cls, buf, offs, idmap = None):
-        key, values = mapped_list.unpack_from(buf, offs, idmap)
-        return cls(zip(key, values))
+        proxy = proxied_dict.unpack_from(buf, offs, idmap)
+        return {k: v for k, v in proxy.iteritems()}
+
+
+def _upsize(x):
+    "Round up X to the next power of 2"
+    for p2 in (1, 2, 4, 8, 16, 32, 64):
+        x |= x >> p2
+    return x + 1
+
+
+def _hash_rotl(code, nbits):
+    return (code << nbits) | (code >> (32 - nbits))
+
+
+def _mix_hash(code1, code2):
+    return _hash_rotl(code1, 5) ^ code2
+
+
+def _stable_hash(key):
+    if isinstance(key, basestring):
+        hval = xxhash.xxh64(key).intdigest()
+    elif isinstance(key, int):
+        hval = key
+    elif isinstance(key, float):
+        mant, expo = math.frexp(key)
+        hval = _mix_hash(expo, int(mant * 0xffffffffffff))
+    elif isinstance(key, (tuple, proxied_tuple, frozenset)):
+        hval = xxhash.xxh64(str(type(key))).intdigest()
+        for value in key:
+            hval = _mix_hash(hval, _stable_hash(value))
+    else:
+        raise TypeError("don't know how to hash key of type %s" % type(key))
+
+    # Truncating the hash value to 48 bits should have no practical impact
+    # on keys' destribution, and it helps us catch subtle overflow bugs
+    return hval & 0xffffffffffff
+
+
+def _enum_keys(obj):
+    idx = 0
+    for key in obj.iterkeys():
+        yield key, idx
+        idx += 1
+
+
+class BufferIO(object):
+    """
+    Gimmick object. Its whole purpose is to implement
+    the protocol expected by the IdMapper classes.
+    """
+
+    def __init__(self, buf, offs):
+        self.buf = buf
+        self.offs = offs
+        self.pos = 0
+
+    def write(self, value):
+        size = len(value)
+        offs = self.offs + self.pos
+        self.buf[offs:offs + size] = value
+        self.pos += size
+        return size
+
+    def tell(self):
+        return self.pos
+
+    def flush(self):
+        pass
+
+    def seek(self, pos):
+        self.pos = pos
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    cython.declare(objmapper=object, vlist=proxied_list)
+
+    def __init__(self, objmapper, vlist):
+        self.objmapper = objmapper
+        self.vlist = vlist
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        ipos = offs
+        offs += 8   # Reserve space for value-list offset.
+        iobuf = BufferIO(buf, offs)
+        mapper_size = ObjectIdMapper.build(_enum_keys(obj), iobuf, map_file=False)
+        offs += mapper_size
+        struct.pack_into("=Q", buf, ipos, mapper_size)
+        return offs + proxied_list.pack_into(obj.values(), buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        mapper_size, = struct.unpack_from("=Q", buf, offs)
```

Comment:
Do as all other classes do, put the header Struct as a class attribute and reuse it